[FLINK-22714][table-planner-blink] Simplify window TVF to a simple window assigner if successor node is WindowRank or WindowJoin.

beyond1920 opened the pull request:
## What is the purpose of the change

In theory, `StreamPhysicalWindowTableFunction` should works as a window assigner and emits at ends of window. However, `StreamPhysicalWindowTableFunction` could be simplified to a simple window assigner, and emits per record if it's successor node is one of `WindowRank`/`WindowJoin`/`WindowAggregate` with `WindowAttachedWindowingStrategy`, because those nodes would assign input records to windows have been assigned in input nodes, and trigger emit at ends of window.

Besides, we should add a limit that window is based on rowtime, because it has syntax problem if window is assigned based on the proc-time of previous operator.

The pr introduce a rule to simplify window-tvf.

## Brief change log

  - Introduce `SimplifyWindowTableFunctionRule`
  - Add a flag in `StreamPhysicalWindowTableFunction` to specify whether it is a simplify window tvf.

## Verifying this change

  - existed UT in `WindowRankTest` and `WindowJoinTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

zjuwangg commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/SimplifyWindowTableFunctionRule.java:
```diff
@@ -0,0 +1,355 @@
+    }
+
+    private boolean needSimplify(StreamPhysicalWindowTableFunction windowTVF) {
+        return windowTVF.windowing().isRowtime() && !windowTVF.emitPerRecord();
```
```suggestion
        return windowTVF.windowing().isRowtime() && windowTVF.emitPerRecord();
```
Do I understand right?👆

beyond1920 replied:
using !windowTVF.emitPerRecord() to exclude those window table function nodes which are already emit per record.  Because there is no need to create new nodes for those window table function nodes.
I could add some comments here to explain logic.

zjuwangg replied:
Thanks for your explanation. I got your meaning.

beyond1920 reviewed (commented):
@zjuwangg Thanks for your review. In fact,

zjuwangg commented:
LGTM. cc @godfreyhe to have another review.

godfreyhe reviewed (commented):
Thanks for the contribution @beyond1920 , I left a few comments.
btw, the description has some typos: `and emits per record if it's successor node is one of WindowRank/WindowJoin/WindowAggregate with WindowAttachedWindowingStrategy`.  `WindowAggregate` do not support this ?

beyond1920 commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/SimplifyWindowTableFunctionRule.java:
```diff
@@ -0,0 +1,355 @@
+        } else if (node instanceof StreamPhysicalWindowJoin) {
+            RelNode newLeft;
+            RelNode newRight;
+            if (call.rule == WITH_WINDOW_JOIN) {
```
Yes, the rule does not support `WindowAggregate` because for window table function + window Aggregate would be matched by `PullUpWindowTableFunctionIntoWindowAggregateRule`.

beyond1920 replied:
Because WindowAggregate could handle inputs with `WindowAttachedWindowingStrategy` and `TimeAttributeWindowingStrategy` which means it could handle watermark assigner itself while `WindowRank` and `Windowjoin` can't.

beyond1920 commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecWindowTableFunction.java:
```diff
@@ -39,15 +39,18 @@
 
-    private final WindowingStrategy windowingStrategy;
+    private final TimeAttributeWindowingStrategy windowingStrategy;
+    private final Boolean emitPerRecord;
```
done

beyond1920 commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/FlinkStreamRuleSets.scala:
```diff
@@ -486,7 +486,13 @@ object FlinkStreamRuleSets {
-  )
+    TwoStageOptimizedWindowAggregateRule.INSTANCE,
+    // simplify window tvf
+    SimplifyWindowTableFunctionRule.WITH_CALC_WINDOW_RANK,
+    SimplifyWindowTableFunctionRule.WITH_WINDOW_RANK,
+    SimplifyWindowTableFunctionRule.WITH_LEFT_RIGHT_CALC_WINDOW_JOIN,
+    SimplifyWindowTableFunctionRule.WITH_LEFT_CALC_WINDOW_JOIN,
+    SimplifyWindowTableFunctionRule.WITH_RIGHT_CALC_WINDOW_JOIN,
+    SimplifyWindowTableFunctionRule.WITH_WINDOW_JOIN)
```
I didn't put into `PHYSICAL_OPT_RULES ` because the rules could not work well in Cost-based-optimization phase.
Because equivalent node didn't have less cost.

beyond1920 commented:
@godfreyhe Thank for your review. I have updated the pr:
1. split the `SimplifyWindowTableFunctionRule` into two subclass.
2. Add json ser/de for `StreamExecWindowTableFunction`, and add `WindowTableFunctionJsonPlanTest` to cover it.

zjuwangg commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecWindowTableFunction.java:
```diff
@@ -21,33 +21,70 @@
-        implements StreamExecNode<RowData> {
+        implements StreamExecNode<RowData>, SingleTransformationTranslator<RowData> {
+
+    public static final String FIELD_NAME_WINDOWING = "windowing";
```
```suggestion
    public static final String FIELD_NAME_WINDOWING = "windowingStrategy";
```

beyond1920 replied:
use `windowing` because keep consistency with WindowAggregate, WindowRank, WindowJoin

zjuwangg replied:
Make sense! Thx

beyond1920 commented:
@godfreyhe I've split `SimplyWindowTableFunctionRule` into multiple subclass, please have a look.
Thanks a lot.

godfreyhe commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/SimplifyWindowTableFunctionRules.java:
```diff
@@ -0,0 +1,317 @@
+        StreamPhysicalWindowRank windowRank = call.rel(0);
+        List<RelNode> rels = call.getRelList();
+        RelNode newTree = rebuild(rels);
+        call.getPlanner().prune(windowRank);
```
unnecessary code

godfreyhe commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/SimplifyWindowTableFunctionRules.java:
```diff
@@ -0,0 +1,317 @@
+
+    @Override
+    StreamPhysicalWindowTableFunction getLeftWindowTVF(RelOptRuleCall call) {
+        return (StreamPhysicalWindowTableFunction) call.rel(2);
```
unnecessary cast

godfreyhe commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/SimplifyWindowTableFunctionRules.java:
```diff
@@ -0,0 +1,317 @@
+
+    @Override
+    StreamPhysicalWindowTableFunction getLeftWindowTVF(RelOptRuleCall call) {
+        return (StreamPhysicalWindowTableFunction) call.rel(3);
```
ditto

godfreyhe commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/SimplifyWindowTableFunctionRules.java:
```diff
@@ -0,0 +1,317 @@
+
+    @Override
+    StreamPhysicalWindowTableFunction getLeftWindowTVF(RelOptRuleCall call) {
+        return (StreamPhysicalWindowTableFunction) call.rel(2);
```
ditto

godfreyhe reviewed (commented):
Thanks for the quick update, LGTM overall, I left a few minor comments

btw, it's better rebase the latest master

beyond1920 commented:
@godfreyhe Thanks for your review, I've address all your comments, please have another look, thanks.

godfreyhe reviewed (approved):
Thanks for the update, LGTM
